packages/theme-language-server-common/src/TypeSystem.ts

```diff
@@ -845,6 +931,101 @@ function settingReturnType(setting: InputSetting): ObjectEntry['return_type'] {
   }
 }
 
+const METAFIELD_TYPE_TO_TYPE = Object.freeze({
+  single_line_text_field: String,
+  multi_line_text_field: String,
+  url_reference: String,
+  date: String,
+  date_time: String,
+  number_integer: 'number',
+  number_decimal: 'number',
+  product_reference: 'product',
+  collection_reference: 'collection',
+  variant_reference: 'variant',
+  page_reference: 'page',
+  boolean: 'boolean',
+  color: 'color',
+  weight: 'measurement',
+  volume: 'measurement',
+  dimension: 'measurement',
+  rating: 'rating',
+  money: 'money',
+  json: Untyped,
+  metaobject_reference: 'metaobject',
+  mixed_reference: Untyped,
+  rich_text_field: Untyped,
+  file_reference: Untyped,
+});
+
+const REFERENCE_TYPE_METAFIELDS = Object.entries(METAFIELD_TYPE_TO_TYPE)
+  .filter(([metafieldType, _type]) => metafieldType.endsWith('_reference'))
+  .map(([_metafieldType, type]) => type);
+
+function metafieldReturnType(metafieldType: string): ObjectEntry['return_type'] {
+  let isArray = metafieldType.startsWith('list.');
+
+  if (isArray) {
+    metafieldType = metafieldType.split('.')[1];
+  }
+
+  let type = 'metafield_' + ((METAFIELD_TYPE_TO_TYPE as any)[metafieldType] ?? Untyped);
+
+  if (isArray) {
+    return [{ type: `${type}_array`, name: '' }];
+  }
+
+  return [{ type: type, name: '' }];
+}
+
+// The default `metafield` type has an untyped `value` property.
+// We need to create new metafield types with the labels `metafield_x` and `metafield_x_array`
+// where x is the type of metafield inside the `value` property. The metafields ending with `x_array`
+// is where the value is an array of type x.
+const customMetafieldTypeEntries = memo((baseMetafieldEntry: ObjectEntry) => {
+  if (!baseMetafieldEntry) return {} as ObjectMap;
+
+  return [
+    ...new Set([...Object.values(METAFIELD_TYPE_TO_TYPE), ...FETCHED_METAFIELD_CATEGORIES]),
+  ].reduce((map, type) => {
+    {
+      const metafieldEntry = JSON.parse(JSON.stringify(baseMetafieldEntry)); // easy deep clone
```

Comment:
Oh are we doing it already. I'll just dm you to verify :D

Reply:
Yeah sorry there are three things happening to make the whole thing work. We can talk about it more today, but here are the things:
- `customMetafieldTypeEntries` --> we copy the `metafield` type and create one for each category we support. e.g. `product_metafield`, `collection_metafield`, etc.
- `metafieldDefinitionsObjectMap` --> reads the metafield definitions we fetched and converts it into types for auto-completion
- the for-loop you commented above --> we override existing types like `product` and `collection`, so their `metafields` properties don't return as `untyped`, but instead return `product_metafields` and `collection_metafields` respectively